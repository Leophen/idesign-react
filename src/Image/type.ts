export interface ImageProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 图片地址
   */
  src?: string;
  /**
   * 图片宽度
   * @default 200
   */
  width?: React.CSSProperties['width'];
}

export interface PreviewDialogProps {
  /**
   * 显示隐藏
   * @default false
   */
  visible?: boolean;
  /**
   * 按下 ESC 是否关闭
   * @default true
   */
  closeOnEsc?: boolean;
  /**
   * 图片地址
   */
  image?: string;
  /**
   * x 值
   * @default 0
   */
  x?: number;
  /**
   * y 值
   * @default 0
   */
  y?: number;
  /**
   * 关闭时触发
   */
  onClose?: () => void;
}
