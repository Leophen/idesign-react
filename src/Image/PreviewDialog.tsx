import React, { useEffect, useRef, useState } from 'react';
import './index.scss';
import Icon from '../Icon';
import Transition from '../Transition';
import { PreviewDialogProps } from './type';

const PreviewDialog: React.FC<PreviewDialogProps> = (props) => {
  const {
    visible = false,
    closeOnEsc = true,
    image,
    x = 0,
    y = 0,
    onClose = () => { }
  } = props;

  const dialog = useRef<any>(null)

  const [origin, setOrigin] = useState({ x: `0px`, y: `0px` })

  const handleKeyDown = (e: any) => {
    if (e.key === 'Escape') {
      onClose?.()
    }
  };

  const closePreviewDialog = (e: React.MouseEvent) => {
    e.stopPropagation()
    onClose?.()
    closeOnEsc && document.removeEventListener('keydown', handleKeyDown)
  }

  const [scale, setScale] = useState(1)
  const handleWheel = (e: any) => {
    e.persist()

    origin.x = `${e.clientX - dialog.current.offsetLeft}px`
    origin.y = `${e.clientY - dialog.current.offsetTop}px`
    setOrigin({ ...origin })

    let newScale = scale
    if (e.deltaY > 0) {
      newScale > 0.4 && (newScale -= 0.1)
    } else {
      newScale < 3 && (newScale += 0.1)
    }
    setScale(newScale)
  }

  const [position, setPosition] = useState({ x: '-50%', y: '-50%' })
  const start = useRef({ x: 0, y: 0 })
  const handleDown = (e: any) => {
    e.stopPropagation()
    start.current.x = e.clientX
    start.current.y = e.clientY
    window.addEventListener('mousemove', handleMove)
  }
  const handleMove = (e: any) => {
    position.x = `calc(-50% + ${e.clientX - start.current.x}px)`
    position.y = `calc(-50% + ${e.clientY - start.current.y}px)`
    setPosition({ ...position })
    window.addEventListener('mouseup', handleUp)
  }
  const handleUp = () => {
    position.x = `-50%`
    position.y = `-50%`
    setPosition({ ...position })

    window.removeEventListener('mousemove', handleMove)
    window.removeEventListener('mouseup', handleUp)
  }

  // 打开对话框时禁止背景滚动，对原 overflow 进行备份
  const bodyOverflow = useRef<string>(document.body.style.overflow);

  useEffect(() => {
    if (visible) {
      // 退出键功能
      closeOnEsc && document.addEventListener('keydown', handleKeyDown)

      // 初始缩放位置
      origin.x = `${x - dialog.current.offsetLeft}px`
      origin.y = `${y - dialog.current.offsetTop}px`
      setOrigin({ ...origin })
      setScale(1)

      // 打开对话框时禁止背景滚动
      document.body.style.overflow = 'hidden';
    }
    return () => {
      closeOnEsc && window.removeEventListener('keydown', handleKeyDown)

      // 关闭对话框时恢复背景滚动
      document.body.style.overflow = bodyOverflow.current;
    }
  }, [visible])

  const clickHandleBar = (e: React.MouseEvent) => {
    e.stopPropagation()
  }

  const [rotate, setRotate] = useState(0)
  const clickHandle = (type: 'reScale' | 'scale' | 'full' | 'reRotate' | 'rotate') => {
    if (type === 'reScale') {
      let newScale = scale
      newScale > 0.4 && (newScale -= 0.1)
      setScale(newScale)
      origin.x = `0px`
      origin.y = `0px`
      setOrigin({ ...origin })
    }
    if (type === 'scale') {
      let newScale = scale
      newScale < 3 && (newScale += 0.1)
      setScale(newScale)
      origin.x = `0px`
      origin.y = `0px`
      setOrigin({ ...origin })
    }
    if (type === 'full') {
      let newScale = 1
      setScale(newScale)
    }
    if (type === 'reRotate') {
      let newRotate = rotate - 90
      setRotate(newRotate)
      origin.x = `center`
      origin.y = `center`
      setOrigin({ ...origin })
    }
    if (type === 'rotate') {
      let newRotate = rotate + 90
      setRotate(newRotate)
      origin.x = `center`
      origin.y = `center`
      setOrigin({ ...origin })
    }
  }

  return (
    <div className='i-preview-dialog-wrapper'>
      <Transition
        timeout={200}
        in={visible}
        animation='fade-in'
        key='i-preview-dialog__mask'
      >
        <div className="i-preview-dialog__mask" onMouseDown={closePreviewDialog} onScroll={() => { return }}>
          <div className="i-preview-dialog__close" onMouseDown={closePreviewDialog}>
            <Icon name="Close" size={20} color="#fff" />
          </div>
          <div className="i-preview-dialog__handle" onMouseDown={clickHandleBar}>
            <section>
              <Icon name="SearchSub" size={24} color="#fff" onClick={() => clickHandle('reScale')} />
              <Icon name="SearchPlus" size={24} color="#fff" onClick={() => clickHandle('scale')} />
            </section>
            <Icon name="FullScreen" size={24} color="#fff" onClick={() => clickHandle('full')} />
            <section>
              <Icon name="RefreshLeft" size={24} color="#fff" onClick={() => clickHandle('reRotate')} />
              <Icon name="RefreshRight" size={24} color="#fff" onClick={() => clickHandle('rotate')} />
            </section>
          </div>
          <img
            style={{
              transform: `scale(${scale}) translate(${position.x}, ${position.y}) rotate(${rotate}deg)`,
              transformOrigin: `${origin.x} ${origin.y}`
            }}
            className='i-preview-img'
            draggable={false}
            src={image}
            ref={dialog}
            onWheel={handleWheel}
            onMouseDown={handleDown}
          />
        </div>
      </Transition>
    </div>
  )
};

export default PreviewDialog;
