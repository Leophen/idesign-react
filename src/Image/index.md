---
nav:
  title: 组件
  path: /components
group:
  title: 数据展示组件
  order: 4
order: 6
---

# Image 图片查看器

用于显示及缩放查看的图片组件。

## 基本用法

```tsx
import React from 'react';
import { Image } from 'idesign-react';

const App = () => {
  return (
    <Image src="https://picsum.photos/180/120" />
  );
};

export default App;
```

## 自定义大小

通过 `width` 属性控制图片大小。

```tsx
import React from 'react';
import { Image } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Image width={260} src="https://picsum.photos/180/120" />
      <Image width="260px" src="https://picsum.photos/180/120" />
    </div>
  );
};

export default App;
```

<API />
