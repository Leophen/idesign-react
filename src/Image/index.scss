.i-image {
  position: relative;
  width: 200px;
}

.i-image-img {
  width: 100%;
  height: auto;
  vertical-align: middle;
}

.i-image-mask {
  position: absolute;
  display: flex;
  gap: 8px;
  align-items: center;
  justify-content: center;
  color: #fff;
  font-size: 14px;
  background: rgba(0, 0, 0, 0.5);
  cursor: pointer;
  opacity: 0;
  transition: opacity 0.3s;
  inset: 0;
  &:hover {
    opacity: 1;
  }
}

.i-preview-dialog-wrapper {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 2500;
  width: 100vw;
  height: 100vh;
  pointer-events: none;
}

.i-preview-dialog__mask {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 2500;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.6);
  pointer-events: auto;
}

.i-preview-dialog__close {
  position: absolute;
  top: 40px;
  right: 40px;
  z-index: 2501;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 45px;
  height: 45px;
  background: var(--i-bg-mid);
  border-radius: 50%;
  cursor: pointer;
  transition: all 0.1s ease-out;
  .i-icon {
    margin: 0;
  }
  &:hover {
    background: var(--i-bg-mid-hover);
  }
}

.i-preview-dialog__handle {
  position: absolute;
  bottom: 30px;
  left: 50%;
  z-index: 2502;
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  width: 360px;
  height: 45px;
  padding: 0 20px;
  background: var(--i-bg-mid);
  border-radius: 20px;
  box-shadow: var(--i-shadow);
  transform: translate(-50%);
  backdrop-filter: blur(4px);
  .i-icon {
    cursor: pointer;
    opacity: 0.8;
    transition: all 0.1s ease-out;
    &:hover {
      opacity: 1;
    }
  }
  section {
    display: flex;
    gap: 16px;
    align-items: center;
  }
}

@keyframes iDialogOpen {
  0% {
    transform: scale(0.2) translate(-50%, -50%);
  }
  100% {
    transform: scale(1) translate(-50%, -50%);
  }
}

.i-preview-img {
  position: absolute;
  top: 50%;
  left: 50%;
  z-index: 2501;
  box-sizing: border-box;
  width: -webkit-fill-available;
  max-width: 100%;
  max-height: 100%;
  box-shadow: var(--i-shadow);
  cursor: grab;
  animation: iDialogOpen 0.2s ease-out;
  pointer-events: auto;
}
