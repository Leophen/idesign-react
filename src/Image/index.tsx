import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import PreviewDialog from './PreviewDialog';
import { ImageProps } from './type';

const Image: React.FC<ImageProps> = (props) => {
  const {
    children = '',
    className,
    style,
    src,
    width,
    ...restProps
  } = props;

  const [previewShow, setPreviewShow] = useState(false)

  const iImage = useRef<any>(null)
  const [dialogPosition, setDialogPosition] = useState({ x: 0, y: 0 })
  useEffect(() => {
    const rect = iImage.current.getBoundingClientRect()
    dialogPosition.x = rect.left
    dialogPosition.y = rect.top
    setDialogPosition({ ...dialogPosition })
  }, [])

  return (
    <div
      className={classNames(
        'i-image',
        className
      )}
      style={{ ...style, width }}
      ref={iImage}
      {...restProps}
    >
      <img className='i-image-img' src={src} />
      <div className="i-image-mask" onClick={() => setPreviewShow(true)}>
        <Icon name="View" color="#fff" />
        预览
      </div>
      <PreviewDialog
        visible={previewShow}
        image={src}
        x={dialogPosition.x}
        y={dialogPosition.y}
        onClose={() => setPreviewShow(false)}
      />
    </div>
  );
};

Image.displayName = 'Image';

export default Image;
