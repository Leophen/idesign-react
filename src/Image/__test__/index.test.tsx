import { render } from '@testing-library/react';
import Image from '../index';

describe('Image 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Image src="https://picsum.photos/180/120" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('width', () => {
    const { container } = render(<Image width={260} src="https://picsum.photos/180/120" />);
    expect(container.firstChild).toHaveStyle('width: 260px');
  })
});

