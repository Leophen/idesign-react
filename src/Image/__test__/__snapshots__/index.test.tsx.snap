// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-image"
  >
    <div
      class="i-loading"
    >
      <div
        class="i-loading-spinner"
      >
        <i
          class="i-icon icon-Loading"
          style="font-size: 36px;"
        />
      </div>
      <div
        class="i-loading-info"
      />
    </div>
    <div
      class="i-image-mask"
    >
      <i
        class="i-icon icon-View"
        style="color: rgb(255, 255, 255);"
      />
      预览
    </div>
    <div
      class="i-preview-dialog-wrapper"
    />
  </div>
</DocumentFragment>
`;
